File: online_judges/kattis/rimstyrka.py
class Node:
    def __init__(self):
        self.is_leaf = False
        self.children = [None] * 26
        self.count = 0

class Trie:
    def __init__(self):
        self.root = Node()
        self.max_power = 0

    def get_index(self, c):
        return ord(c) - ord('a')

    def insert(self, word):
        cur = self.root
        cnt  = 0
        for i in range(len(word)-1, -1, -1):
            index = self.get_index(word[i])
            if cur.children[index] is None:
                if cur.count > 0:
                    self.max_power = max(self.max_power, cnt)
                    
                cur.children[index] = Node()
                cur.count += 1
                
            else:
                cnt += 1
            cur = cur.children[index]
        cur.is_leaf = True

        

t = int(raw_input())
trie = Trie()

for _ in range(t):
    trie.insert(raw_input())

print(trie.max_power)

File: online_judges/kattis/rimstyrka.cpp
#include <bits/stdc++.h>

using namespace std;


int ans = 0;
class Trie {
private:
	vector<Trie*> children;
	int node_count;
	bool leaf;

public:

	Trie(): children(26), node_count(0), leaf(false){
	};


	void insert(string word) {
		Trie* cur = this;
		int cnt = 0;
		for(int i = word.size()-1; i >= 0 ; i--) {
		  int c = word[i] - 'a';
		  if(cur->children[c] == NULL) {
			if (cur->node_count > 0) {
				ans = max(ans, cnt);
			}
			cur->children[c] = new Trie();
			cur->node_count++;
		  }
		  else cnt ++;
		  cur = cur->children[c];
		}
		cur->leaf = true;
	}
};

int main(){
	Trie* trie = new Trie();
	int n;
	cin >> n;
	for(int i= 0;i<n;i++){
		string s;
		cin >> s;
		trie->insert(s);
	}
	cout << ans << endl;
	return 0;
}
